man/im_msb.3
.TH IM_MSB 3 "06 May 2006"
.SH NAME
 im_msb, im_msb_band \- Convert to uchar by discarding bits
.SH SYNOPSIS
.nf
.B #include <vips/vips.h>
.sp
.BI "int im_msb( IMAGE " "*in" ", IMAGE " "*out" " );
.br

.BI "int im_msb_band( IMAGE " "*in" ", IMAGE " "*out" ", int " "band" " );
.fi
.SH DESCRIPTION
.B im_msb(3)
converts char, short, or int images (including LABQ coded ones) into unsigned
char images, very quickly, by discarding the lower order bits.  Once scaled to
char, signed values are converted to unsigned by adding 128.
.PP
For a signed short (16 bit) image,
.PP
im_msb( in, out );
.PP
is equivalent to:
.PP
im_lintra( (1.0/256.0), in, 128.0, temp );
.br
im_clip2fmt( temp, out, IM_BANDFMT_UCHAR );
.PP
but much faster.
.PP
For any image which uses the whole range of values for its band format,
.PP
im_msb( in, out );
.PP
is equivalent to:
.PP
im_scale( in, out );
.PP
but a great deal faster, and without evaluating the input twice.
.PP
.B im_msb_band(3)
is as
.BR im_msb(3) ,
except that all but one of the bands are also discarded.
.PP
im_msb_band( in, out, i );
.PP
is equivalent to:
.PP
im_msb( in, temp );
.br
im_extract_bands( temp, out, i, 1 );
.PP
but again, faster.
.SH RETURN VALUE
The functions returns 0 on success and -1 on error.
.SH SEE ALSO
im_lintra(3), im_scale(3), im_clip(3)
.SH COPYRIGHT
.br
Copyright 2006, The Nottingham Trent University.
.SH AUTHOR
Tom Vajzovic


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/conversion/im_msb.c
/* @(#) Convert a signed or unsigned, char or short or int image into unsigned
 * @(#) char very quickly, by discarding the lower order bits.  Signed values
 * @(#) are converted to unsigned by adding 128.
 * @(#)
 * @(#) int
 * @(#) im_msb(
 * @(#)   IMAGE *in,
 * @(#)   IMAGE *out
 * @(#) );
 * @(#)
 * @(#) As above, but also discard all but the specified band:
 * @(#)
 * @(#) int
 * @(#) im_msb(
 * @(#)   IMAGE *in,
 * @(#)   IMAGE *out,
 * @(#)   int band
 * @(#) );
 *
 * Copyright: 2006, The Nottingham Trent University
 *
 * Author: Tom Vajzovic
 *
 * Written on: 2006-03-13
 * 27/9/06
 * 	- removed extra im_free() in im_copy() fallback
 * 4/10/06
 * 	- removed warning on uchar fallback: it happens a lot with nip2 and
 * 	  isn't very serious
 */

/*

    This file is part of VIPS.

    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

/** HEADERS **/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H */
#include <vips/intl.h>

#include <stdlib.h>

#include <vips/vips.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC */

/** MACROS **/
/* BAND FORMAT TESTS */

#define SIZEOF_BAND(band_fmt) (                                       \
     ( IM_BANDFMT_UCHAR == (band_fmt) ) ? sizeof(unsigned char)       \
 :   ( IM_BANDFMT_CHAR  == (band_fmt) ) ? sizeof(signed char)         \
 :  ( IM_BANDFMT_USHORT == (band_fmt) ) ? sizeof(unsigned short int)  \
 :   ( IM_BANDFMT_SHORT == (band_fmt) ) ? sizeof(signed short int)    \
 :    ( IM_BANDFMT_UINT == (band_fmt) ) ? sizeof(unsigned int)        \
 :     ( IM_BANDFMT_INT == (band_fmt) ) ? sizeof(signed int)          \
 :   ( IM_BANDFMT_FLOAT == (band_fmt) ) ? sizeof(float)               \
 :  ( IM_BANDFMT_DOUBLE == (band_fmt) ) ? sizeof(double)              \
 : ( IM_BANDFMT_COMPLEX == (band_fmt) ) ? ( 2 * sizeof(float) )       \
 :                                        ( 2 * sizeof(double) ) )

#define BANDFMT_SHORT_CHAR(band_fmt)  (  \
       IM_BANDFMT_SHORT == (band_fmt)    \
  ||  IM_BANDFMT_USHORT == (band_fmt)    \
  ||    IM_BANDFMT_CHAR == (band_fmt)    \
  ||   IM_BANDFMT_UCHAR == (band_fmt) )

#define BANDFMT_ANY_INT(band_fmt)    (  \
         IM_BANDFMT_INT == (band_fmt)   \
  ||    IM_BANDFMT_UINT == (band_fmt)   \
  ||     BANDFMT_SHORT_CHAR(band_fmt) )

#define BANDFMT_UNSIGNED(band_fmt)  (   \
        IM_BANDFMT_UINT == (band_fmt)   \
  ||  IM_BANDFMT_USHORT == (band_fmt)   \
  ||   IM_BANDFMT_UCHAR == (band_fmt) )

/* IMAGE TESTS */

#define IM_ANY_INT(im)      BANDFMT_ANY_INT((im)-> BandFmt)
#define IM_UNSIGNED(im)     BANDFMT_UNSIGNED((im)-> BandFmt)
#define IM_UNCODED(im)      ( IM_CODING_NONE == (im)-> Coding )

/** LOCAL FUNCTIONS DECLARATIONS **/
static void byte_select (unsigned char *in, unsigned char *out, int n,
			 size_t * params);
static void byte_select_flip (unsigned char *in, unsigned char *out, int n,
			      size_t * params);
static void msb_labq (unsigned char *in, unsigned char *out, int n);

/** EXPORTED FUNCTIONS **/

int
im_msb (IMAGE * in, IMAGE * out)
{
#define FUNCTION_NAME "im_msb"

  size_t *params;
  im_wrapone_fn func;

#define index   (params[0])
#define width   (params[1])
#define repeat  (params[2])

  if (im_piocheck (in, out))
    return -1;

  /* Stops a used-before-set warning.
   */
  params = NULL;

  if (IM_UNCODED (in))
    {

      if (!IM_ANY_INT (in))
	{
	  im_error (FUNCTION_NAME, _("char, short or int only"));
	  return -1;
	}

      params = IM_ARRAY (out, 3, size_t);

      if (!params)
	return -1;

      width = SIZEOF_BAND (in->BandFmt);

#ifdef WORDS_BIGENDIAN
      index = 0;
#else
      index = width - 1;
#endif

      repeat = in->Bands;

      if (IM_UNSIGNED (in))
	func = (im_wrapone_fn) byte_select;
      else
	func = (im_wrapone_fn) byte_select_flip;

      if (1 == width && (im_wrapone_fn) byte_select == func)
	{
	  return im_copy (in, out);
	}
    }

  else if (IM_CODING_LABQ == in->Coding)

    func = (im_wrapone_fn) msb_labq;

  else
    {
      im_error (FUNCTION_NAME, _("unknown coding"));
      return -1;
    }

  if (im_cp_desc (out, in))
    return -1;

  out->Bbits = sizeof (unsigned char) << 3;
  out->BandFmt = IM_BANDFMT_UCHAR;
  out->Coding = IM_CODING_NONE;

  return im_wrapone (in, out, func, (void *) params, NULL);

#undef index
#undef width
#undef repeat

#undef FUNCTION_NAME
}

int
im_msb_band (IMAGE * in, IMAGE * out, int band)
{
#define FUNCTION_NAME "im_msb_band"

  size_t *params;
  im_wrapone_fn func;

#define index   (params[0])
#define width   (params[1])
#define repeat  (params[2])

  if (band < 0)
    {
      im_error (FUNCTION_NAME, _("bad arguments"));
      return -1;
    }

  if (im_piocheck (in, out))
    return -1;

  params = IM_ARRAY (out, 3, size_t);

  if (!params)
    return -1;

  if (IM_UNCODED (in))
    {

      if (!IM_ANY_INT (in))
	{
	  im_error (FUNCTION_NAME, _("char, short or int only"));
	  return -1;
	}

      if (band >= in->Bands)
	{
	  im_error (FUNCTION_NAME, _("image does not have that many bands"));
	  return -1;
	}

      width = SIZEOF_BAND (in->BandFmt);

#ifdef WORDS_BIGENDIAN
      index = width * band;
#else
      index = (width * (band + 1)) - 1;
#endif

      width *= in->Bands;
      repeat = 1;

      if (IM_UNSIGNED (in))
	func = (im_wrapone_fn) byte_select;
      else
	func = (im_wrapone_fn) byte_select_flip;
    }

  else if (IM_CODING_LABQ == in->Coding)
    {

      if (band > 2)
	{
	  im_error (FUNCTION_NAME, _("image does not have that many bands"));
	  return -1;
	}
      width = 4;
      repeat = 1;
      index = band;

      if (band)
	func = (im_wrapone_fn) byte_select_flip;
      else
	func = (im_wrapone_fn) byte_select;
    }
  else
    {
      im_error (FUNCTION_NAME, _("unknown coding"));
      return -1;
    }

  if (im_cp_desc (out, in))
    return -1;

  out->Bands = 1;
  out->Bbits = sizeof (unsigned char) << 3;
  out->BandFmt = IM_BANDFMT_UCHAR;
  out->Coding = IM_CODING_NONE;

  return im_wrapone (in, out, func, (void *) params, NULL);

#undef index
#undef width
#undef repeat

#undef FUNCTION_NAME
}

/** LOCAL FUNCTIONS DEFINITIONS **/
static void
byte_select (unsigned char *in, unsigned char *out, int n, size_t * params)
{
#define index   (params[0])
#define width   (params[1])
#define repeat  (params[2])

  unsigned char *stop = out + n * repeat;

  for (in += index; out < stop; in += width, ++out)
    *out = *in;

#undef index
#undef width
#undef repeat
}

static void
byte_select_flip (unsigned char *in, unsigned char *out, int n,
		  size_t * params)
{
#define index   (params[0])
#define width   (params[1])
#define repeat  (params[2])

  unsigned char *stop = out + n * repeat;

  for (in += index; out < stop; in += width, ++out)
    *out = 0x80 ^ *in;

#undef index
#undef width
#undef repeat
}

static void
msb_labq (unsigned char *in, unsigned char *out, int n)
{
  unsigned char *stop = in + (n << 2);

  for (; in < stop; in += 4, out += 3)
    {
      out[0] = in[0];
      out[1] = 0x80 ^ in[1];
      out[2] = 0x80 ^ in[2];
    }
}
